tests/unit/test_hnsw_parallel.cpp

```diff
@@ -466,3 +508,268 @@ TYPED_TEST(HNSWTestParallel, parallelInsertSearch) {
+
+    VecSimIndex_Free(hnsw_index);
+}
+
+TYPED_TEST(HNSWTestParallel, parallelRepairSearch) {
+    size_t n = 10000;
+    size_t k = 10;
+    size_t dim = 32;
+
+    HNSWParams params = {
+        .dim = dim, .metric = VecSimMetric_L2, .initialCapacity = n, .efRuntime = n};
+
+    auto *hnsw_index = this->CastToHNSW(this->CreateNewIndex(params));
+    size_t n_threads = MIN(10, std::thread::hardware_concurrency());
+    // Save the number fo tasks done by thread i in the i-th entry.
+    std::vector<size_t> completed_tasks(n_threads, 0);
+
+    for (size_t i = 0; i < n; i++) {
+        GenerateAndAddVector<TEST_DATA_T>(hnsw_index, dim, i, i);
+    }
+    ASSERT_EQ(VecSimIndex_IndexSize(hnsw_index), n);
+
+    // Queue of repair jobs, each job is represented as {id, level}
+    auto jobQ = std::queue<pair<idType, size_t>>();
+    std::mutex queue_guard;
+
+    for (size_t element_id = 0; element_id < n; element_id += 2) {
+        hnsw_index->markDelete(element_id);
+    }
+    ASSERT_EQ(hnsw_index->getNumMarkedDeleted(), n / 2);
+    // Every deleted node i should have at least 2 connection to repair (to i-1 and i-1), except for
+    // 0 and n-1 that has at least one connection to repair.
+    ASSERT_GE(hnsw_index->checkIntegrity().connection_to_repair, n - 2);
+
+    // Collect all the nodes that require repairment due to the deletions, from top level down.
+    this->CollectRepairJobs(hnsw_index, jobQ);
+
+    auto executeRepairJob = [&](int myID) {
+        queue_guard.lock();
+        while (!jobQ.empty()) {
+            auto job = jobQ.front();
+            jobQ.pop();
+            queue_guard.unlock();
+            hnsw_index->repairNodeConnections(job.first, job.second); // {element_id, level}
+            completed_tasks[myID]++;
+            queue_guard.lock();
+        }
+        queue_guard.unlock();
+    };
+
+    TEST_DATA_T query_val = (TEST_DATA_T)n / 4;
+    std::atomic_int successful_searches(0);
+    auto parallel_knn_search = [&](int myID) {
+        completed_tasks[myID]++;
+        TEST_DATA_T query[dim];
+        GenerateVector<TEST_DATA_T>(query, dim, query_val);
+        auto verify_res = [&](size_t id, double score, size_t res_index) {
+            // We expect to get the results with increasing order of the distance between the
+            // res label and the query val (n/4-1, n/4+1, n/4-3, n/4+3, ...) The score is
+            // the L2 distance between the vectors that correspond the ids.
+            size_t diff_id = std::abs(int(id - query_val));
+            ASSERT_EQ(diff_id, res_index + (1 - res_index % 2));
+            ASSERT_EQ(score, (dim * (diff_id * diff_id)));
+        };
+        runTopKSearchTest(hnsw_index, query, k, verify_res);
+        successful_searches++;
+    };
+
+    std::thread thread_objs[n_threads];
+    for (size_t i = 0; i < n_threads / 2; i++) {
+        thread_objs[i] = std::thread(executeRepairJob, i);
+    }
+
+    // Start running queries after half of the job are executed.
```

Comment:
What's the advantage of doing so?

Reply:
consider having the querying threads run in a loop until signaled (by a variable) to stop, and make them run the entire runtime of the repairing threads